--- src/Dockerfile
+++ src/Dockerfile
@@ -3,21 +3,12 @@
 ARG VERSION=1.23.0
 ARG PREFIX=/w64devkit
 ARG BINUTILS_VERSION=2.42
-ARG BUSYBOX_VERSION=FRP-5301-gda71f7c57
-ARG CTAGS_VERSION=6.0.0
-ARG EXPAT_VERSION=2.6.2
 ARG GCC_VERSION=14.1.0
-ARG GDB_VERSION=14.2
 ARG GMP_VERSION=6.3.0
-ARG LIBICONV_VERSION=1.17
 ARG MAKE_VERSION=4.4.1
 ARG MINGW_VERSION=12.0.0
 ARG MPC_VERSION=1.3.1
 ARG MPFR_VERSION=4.2.1
-ARG NASM_VERSION=2.15.05
-ARG PDCURSES_VERSION=3.9
-ARG CPPCHECK_VERSION=2.10
-ARG VIM_VERSION=9.0
 
 RUN apt-get update && apt-get install --yes --no-install-recommends \
   build-essential curl libgmp-dev libmpc-dev libmpfr-dev m4 zip
@@ -27,50 +18,31 @@
 RUN curl --insecure --location --remote-name-all --remote-header-name \
     https://ftp.gnu.org/gnu/binutils/binutils-$BINUTILS_VERSION.tar.xz \
     https://ftp.gnu.org/gnu/gcc/gcc-$GCC_VERSION/gcc-$GCC_VERSION.tar.xz \
-    https://ftp.gnu.org/gnu/gdb/gdb-$GDB_VERSION.tar.xz \
-    https://fossies.org/linux/www/expat-$EXPAT_VERSION.tar.xz \
     https://ftp.gnu.org/gnu/gmp/gmp-$GMP_VERSION.tar.xz \
     https://ftp.gnu.org/gnu/mpc/mpc-$MPC_VERSION.tar.gz \
     https://ftp.gnu.org/gnu/mpfr/mpfr-$MPFR_VERSION.tar.xz \
-    https://ftp.gnu.org/gnu/make/make-$MAKE_VERSION.tar.gz \
-    https://ftp.gnu.org/gnu/libiconv/libiconv-$LIBICONV_VERSION.tar.gz \
-    https://frippery.org/files/busybox/busybox-w32-$BUSYBOX_VERSION.tgz \
-    http://ftp.vim.org/pub/vim/unix/vim-$VIM_VERSION.tar.bz2 \
-    https://www.nasm.us/pub/nasm/releasebuilds/$NASM_VERSION/nasm-$NASM_VERSION.tar.xz \
-    https://github.com/universal-ctags/ctags/archive/refs/tags/v$CTAGS_VERSION.tar.gz \
-    https://downloads.sourceforge.net/project/mingw-w64/mingw-w64/mingw-w64-release/mingw-w64-v$MINGW_VERSION.tar.bz2 \
-    https://downloads.sourceforge.net/project/pdcurses/pdcurses/$PDCURSES_VERSION/PDCurses-$PDCURSES_VERSION.tar.gz \
-    https://github.com/danmar/cppcheck/archive/$CPPCHECK_VERSION.tar.gz
+    https://downloads.sourceforge.net/project/mingw-w64/mingw-w64/mingw-w64-release/mingw-w64-v$MINGW_VERSION.tar.bz2
 COPY src/SHA256SUMS $PREFIX/src/
 RUN sha256sum -c $PREFIX/src/SHA256SUMS \
  && tar xJf binutils-$BINUTILS_VERSION.tar.xz \
- && tar xzf busybox-w32-$BUSYBOX_VERSION.tgz \
- && tar xzf ctags-$CTAGS_VERSION.tar.gz \
  && tar xJf gcc-$GCC_VERSION.tar.xz \
- && tar xJf gdb-$GDB_VERSION.tar.xz \
- && tar xJf expat-$EXPAT_VERSION.tar.xz \
- && tar xzf libiconv-$LIBICONV_VERSION.tar.gz \
  && tar xJf gmp-$GMP_VERSION.tar.xz \
  && tar xzf mpc-$MPC_VERSION.tar.gz \
  && tar xJf mpfr-$MPFR_VERSION.tar.xz \
- && tar xzf make-$MAKE_VERSION.tar.gz \
- && tar xjf mingw-w64-v$MINGW_VERSION.tar.bz2 \
- && tar xzf PDCurses-$PDCURSES_VERSION.tar.gz \
- && tar xJf nasm-$NASM_VERSION.tar.xz \
- && tar xjf vim-$VIM_VERSION.tar.bz2 \
- && tar xzf cppcheck-$CPPCHECK_VERSION.tar.gz
-COPY src/w64devkit.c src/w64devkit.ico src/libmemory.c src/libchkstk.S \
-     src/alias.c src/debugbreak.c src/pkg-config.c src/vc++filt.c \
-     src/peports.c $PREFIX/src/
+ && tar xjf mingw-w64-v$MINGW_VERSION.tar.bz2
+COPY src/libmemory.c src/libchkstk.S src/vc++filt.c src/peports.c \
+     $PREFIX/src/
 
 ARG ARCH=x86_64-w64-mingw32
 
 # Build cross-compiler
 
+COPY minimal-pe-header.patch /binutils-$BINUTILS_VERSION/
 WORKDIR /binutils-$BINUTILS_VERSION
 COPY src/binutils-*.patch $PREFIX/src/
 RUN sed -ri 's/(static bool insert_timestamp = )/\1!/' ld/emultempl/pe*.em \
- && cat $PREFIX/src/binutils-*.patch | patch -p1
+ && cat $PREFIX/src/binutils-*.patch | patch -p1 \
+ && patch -p 0 -u -i minimal-pe-header.patch
 WORKDIR /x-binutils
 RUN /binutils-$BINUTILS_VERSION/configure \
         --prefix=/bootstrap \
@@ -288,31 +260,7 @@
  && make -j$(nproc) \
  && make install \
  && rm -rf $PREFIX/$ARCH/bin/ $PREFIX/bin/$ARCH-* \
-        $PREFIX/bin/ld.bfd.exe $PREFIX/bin/c++.exe \
- && $ARCH-gcc -DEXE=g++.exe -DCMD=c++ \
-        -Os -fno-asynchronous-unwind-tables \
-        -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/c++.exe \
-        $PREFIX/src/alias.c -lkernel32
-
-# Create various tool aliases
-RUN $ARCH-gcc -DEXE=gcc.exe -DCMD=cc \
-        -Os -fno-asynchronous-unwind-tables -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/cc.exe $PREFIX/src/alias.c -lkernel32 \
- && $ARCH-gcc -DEXE=gcc.exe -DCMD="cc -std=c99" \
-        -Os -fno-asynchronous-unwind-tables -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/c99.exe $PREFIX/src/alias.c -lkernel32 \
- && $ARCH-gcc -DEXE=gcc.exe -DCMD="cc -ansi" \
-        -Os -fno-asynchronous-unwind-tables -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/c89.exe $PREFIX/src/alias.c -lkernel32 \
- && printf '%s\n' addr2line ar as c++filt cpp dlltool dllwrap elfedit g++ \
-      gcc gcc-ar gcc-nm gcc-ranlib gcov gcov-dump gcov-tool ld nm objcopy \
-      objdump ranlib readelf size strings strip windmc windres \
-    | xargs -I{} -P$(nproc) \
-          $ARCH-gcc -DEXE={}.exe -DCMD=$ARCH-{} \
-            -Os -fno-asynchronous-unwind-tables \
-            -Wl,--gc-sections -s -nostdlib \
-            -o $PREFIX/bin/$ARCH-{}.exe $PREFIX/src/alias.c -lkernel32
+        $PREFIX/bin/ld.bfd.exe $PREFIX/bin/c++.exe
 
 # Build some extra development tools
 
@@ -326,191 +274,19 @@
  && make -j$(nproc) \
  && cp gendef.exe $PREFIX/bin/
 
-WORKDIR /expat
-RUN /expat-$EXPAT_VERSION/configure \
-        --prefix=/deps \
-        --host=$ARCH \
-        --disable-shared \
-        --without-docbook \
-        --without-examples \
-        --without-tests \
-        CFLAGS="-Os" \
-        LDFLAGS="-s" \
- && make -j$(nproc) \
- && make install
-
-WORKDIR /PDCurses-$PDCURSES_VERSION
-RUN make -j$(nproc) -C wincon \
-        CC=$ARCH-gcc AR=$ARCH-ar CFLAGS="-I.. -Os -DPDC_WIDE" pdcurses.a \
- && cp wincon/pdcurses.a /deps/lib/libcurses.a \
- && cp curses.h /deps/include
-
-WORKDIR /libiconv
-RUN /libiconv-$LIBICONV_VERSION/configure \
-        --prefix=/deps \
-        --host=$ARCH \
-        --disable-nls \
-        --disable-shared \
-        CFLAGS="-Os" \
-        LDFLAGS="-s" \
- && make -j$(nproc) \
- && make install
-
-WORKDIR /gdb
-COPY src/gdb-*.patch $PREFIX/src/
-RUN cat $PREFIX/src/gdb-*.patch | patch -d/gdb-$GDB_VERSION -p1 \
- && sed -i 's/quiet = 0/quiet = 1/' /gdb-$GDB_VERSION/gdb/main.c \
- && /gdb-$GDB_VERSION/configure \
-        --host=$ARCH \
-        --enable-tui \
-        CFLAGS="-Os -D__MINGW_USE_VC2005_COMPAT -DPDC_WIDE -I/deps/include" \
-        CXXFLAGS="-Os -D__MINGW_USE_VC2005_COMPAT -DPDC_WIDE -I/deps/include" \
-        LDFLAGS="-s -L/deps/lib" \
- && make MAKEINFO=true -j$(nproc) \
- && cp gdb/.libs/gdb.exe gdbserver/gdbserver.exe $PREFIX/bin/
-
-WORKDIR /make
-RUN /make-$MAKE_VERSION/configure \
-        --host=$ARCH \
-        --disable-nls \
-        CFLAGS="-Os" \
-        LDFLAGS="-s" \
- && make -j$(nproc) \
- && cp make.exe $PREFIX/bin/ \
- && $ARCH-gcc -DEXE=make.exe -DCMD=make \
-        -Os -fno-asynchronous-unwind-tables \
-        -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/mingw32-make.exe $PREFIX/src/alias.c -lkernel32
-
-WORKDIR /busybox-w32
-COPY src/busybox-* $PREFIX/src/
-RUN cat $PREFIX/src/busybox-*.patch | patch -p1 \
- && make mingw64_defconfig \
- && sed -ri 's/^(CONFIG_AR)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_ASCII)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_DPKG\w*)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_FTP\w*)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_LINK)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_MAN)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_MAKE)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_PDPMAKE)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_RPM\w*)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_STRINGS)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_TEST2)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_TSORT)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_UNLINK)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_VI)=y/\1=n/' .config \
- && sed -ri 's/^(CONFIG_XXD)=y/\1=n/' .config \
- && make -j$(nproc) CROSS_COMPILE=$ARCH- \
-    CONFIG_EXTRA_CFLAGS="-D_WIN32_WINNT=0x502" \
- && cp busybox.exe $PREFIX/bin/
-
-# Create BusyBox command aliases (like "busybox --install")
-RUN $ARCH-gcc -Os -fno-asynchronous-unwind-tables -Wl,--gc-sections -s \
-      -nostdlib -o alias.exe $PREFIX/src/busybox-alias.c -lkernel32 \
- && printf '%s\n' arch ash awk base32 base64 basename bash bc bunzip2 bzcat \
-      bzip2 cal cat chattr chmod cksum clear cmp comm cp cpio crc32 cut date \
-      dc dd df diff dirname dos2unix du echo ed egrep env expand expr factor \
-      false fgrep find fold free fsync getopt grep groups gunzip gzip hd \
-      head hexdump httpd iconv id inotifyd install ipcalc jn kill killall \
-      lash less ln logname ls lsattr lzcat lzma lzop lzopcat md5sum mkdir \
-      mktemp mv nc nl nproc od paste patch pgrep pidof pipe_progress pkill \
-      printenv printf ps pwd readlink realpath reset rev rm rmdir sed seq sh \
-      sha1sum sha256sum sha3sum sha512sum shred shuf sleep sort split \
-      ssl_client stat su sum sync tac tail tar tee test time timeout touch \
-      tr true truncate ts ttysize uname uncompress unexpand uniq unix2dos \
-      unlzma unlzop unxz unzip uptime usleep uudecode uuencode watch \
-      wc wget which whoami whois xargs xz xzcat yes zcat \
-    | xargs -I{} cp alias.exe $PREFIX/bin/{}.exe
-
-# TODO: Either somehow use $VIM_VERSION or normalize the workdir
-WORKDIR /vim90/src
-RUN ARCH= make -j$(nproc) -f Make_ming.mak \
-        OPTIMIZE=SIZE STATIC_STDCPLUS=yes HAS_GCC_EH=no \
-        UNDER_CYGWIN=yes CROSS=yes CROSS_COMPILE=$ARCH- \
-        FEATURES=HUGE VIMDLL=yes NETBEANS=no WINVER=0x0501 \
- && $ARCH-strip vimrun.exe \
- && rm -rf ../runtime/tutor/tutor.* \
- && cp -r ../runtime $PREFIX/share/vim \
- && cp vimrun.exe gvim.exe vim.exe *.dll $PREFIX/share/vim/ \
- && cp xxd/xxd.exe $PREFIX/bin \
- && printf '@set SHELL=\r\n@start "" "%%~dp0/../share/vim/gvim.exe" %%*\r\n' \
-        >$PREFIX/bin/gvim.bat \
- && printf '@set SHELL=\r\n@"%%~dp0/../share/vim/vim.exe" %%*\r\n' \
-        >$PREFIX/bin/vim.bat \
- && printf '@set SHELL=\r\n@"%%~dp0/../share/vim/vim.exe" %%*\r\n' \
-        >$PREFIX/bin/vi.bat \
- && printf '@vim -N -u NONE "+read %s" "+write" "%s"\r\n' \
-        '$VIMRUNTIME/tutor/tutor' '%TMP%/tutor%RANDOM%' \
-        >$PREFIX/bin/vimtutor.bat
-
-# NOTE: nasm's configure script is broken, so no out-of-source build
-WORKDIR /nasm-$NASM_VERSION
-RUN ./configure \
-        --host=$ARCH \
-        CFLAGS="-Os" \
-        LDFLAGS="-s" \
- && make -j$(nproc) \
- && cp nasm.exe ndisasm.exe $PREFIX/bin
-
-WORKDIR /ctags-$CTAGS_VERSION
-RUN sed -i /RT_MANIFEST/d win32/ctags.rc \
- && make -j$(nproc) -f mk_mingw.mak CC=gcc packcc.exe \
- && make -j$(nproc) -f mk_mingw.mak \
-        CC=$ARCH-gcc WINDRES=$ARCH-windres \
-        OPT= CFLAGS=-Os LDFLAGS=-s \
- && cp ctags.exe $PREFIX/bin/
-
-WORKDIR /cppcheck-$CPPCHECK_VERSION
-COPY src/cppcheck.mak src/cppcheck-*.patch $PREFIX/src/
-RUN cat $PREFIX/src/cppcheck-*.patch | patch -p1 \
- && make -f $PREFIX/src/cppcheck.mak -j$(nproc) CXX=$ARCH-g++ \
- && mkdir $PREFIX/share/cppcheck/ \
- && cp -r cppcheck.exe cfg/ $PREFIX/share/cppcheck \
- && $ARCH-gcc -DEXE=../share/cppcheck/cppcheck.exe -DCMD=cppcheck \
-        -Os -fno-asynchronous-unwind-tables -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/cppcheck.exe \
-        $PREFIX/src/alias.c -lkernel32
-
 # Pack up a release
 
 WORKDIR /
 RUN rm -rf $PREFIX/share/man/ $PREFIX/share/info/ $PREFIX/share/gcc-* \
- && rm -rf $PREFIX/lib/*.a $PREFIX/lib/*.la $PREFIX/include/*.h
-COPY README.md Dockerfile src/w64devkit.ini $PREFIX/
-RUN printf "id ICON \"$PREFIX/src/w64devkit.ico\"" >w64devkit.rc \
- && $ARCH-windres -o w64devkit.o w64devkit.rc \
- && $ARCH-gcc -DVERSION=$VERSION -nostdlib -fno-asynchronous-unwind-tables \
-        -fno-builtin -Wl,--gc-sections -s -o $PREFIX/w64devkit.exe \
-        $PREFIX/src/w64devkit.c w64devkit.o -lkernel32 -luser32 \
- && $ARCH-gcc \
-        -Os -fno-asynchronous-unwind-tables \
-        -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/debugbreak.exe $PREFIX/src/debugbreak.c \
-        -lkernel32 \
- && $ARCH-gcc \
-        -Os -fno-asynchronous-unwind-tables -fno-builtin -Wl,--gc-sections \
-        -s -nostdlib -DPKG_CONFIG_PREFIX="\"/$ARCH\"" \
-        -o $PREFIX/bin/pkg-config.exe $PREFIX/src/pkg-config.c \
-        -lkernel32 \
- && $ARCH-gcc \
+ && rm -rf $PREFIX/lib/*.a $PREFIX/lib/*.la $PREFIX/include $PREFIX/share
+RUN $ARCH-gcc \
         -Os -fno-asynchronous-unwind-tables -fno-builtin -Wl,--gc-sections \
         -s -nostdlib -o $PREFIX/bin/vc++filt.exe $PREFIX/src/vc++filt.c \
         -lkernel32 -lshell32 -ldbghelp \
  && $ARCH-gcc \
         -Os -fno-asynchronous-unwind-tables -fno-builtin -Wl,--gc-sections \
         -s -nostdlib -o $PREFIX/bin/peports.exe $PREFIX/src/peports.c \
-        -lkernel32 -lshell32 \
- && $ARCH-gcc -DEXE=pkg-config.exe -DCMD=pkg-config \
-        -Os -fno-asynchronous-unwind-tables -Wl,--gc-sections -s -nostdlib \
-        -o $PREFIX/bin/$ARCH-pkg-config.exe $PREFIX/src/alias.c -lkernel32 \
- && mkdir -p $PREFIX/$ARCH/lib/pkgconfig \
- && cp /mingw-w64-v$MINGW_VERSION/COPYING.MinGW-w64-runtime/COPYING.MinGW-w64-runtime.txt \
-        $PREFIX/ \
- && printf "\n===========\nwinpthreads\n===========\n\n" \
-        >>$PREFIX/COPYING.MinGW-w64-runtime.txt . \
- && cat /mingw-w64-v$MINGW_VERSION/mingw-w64-libraries/winpthreads/COPYING \
-        >>$PREFIX/COPYING.MinGW-w64-runtime.txt \
- && echo $VERSION >$PREFIX/VERSION.txt
+        -lkernel32 -lshell32
+RUN rm -rf $PREFIX/src
 ENV PREFIX=${PREFIX}
 CMD zip -q9Xr - $PREFIX
--- src/src/SHA256SUMS
+++ src/src/SHA256SUMS
@@ -1,16 +1,6 @@
 f6e4d41fd5fc778b06b7891457b3620da5ecea1006c6a4a41ae998109f85a800  binutils-2.42.tar.xz
-ebdea8aa03d9d457075f3ff5ce4dc04239cd4b9ba6439bb701f285233c960923  busybox-w32-FRP-5301-gda71f7c57.tgz
-785dcbf711048dfe43ae920b6eff2eeebb4a096e88188a40e173ca4c030f57c3  cppcheck-2.10.tar.gz
-71229a73f25529c9e3dabb2cb7310c55405d31caee8e8a9ab5c71b2406d4005a  ctags-6.0.0.tar.gz
-ee14b4c5d8908b1bec37ad937607eab183d4d9806a08adee472c3c3121d27364  expat-2.6.2.tar.xz
 e283c654987afe3de9d8080bc0bd79534b5ca0d681a73a11ff2b5d3767426840  gcc-14.1.0.tar.xz
-2d4dd8061d8ded12b6c63f55e45344881e8226105f4d2a9b234040efa5ce7772  gdb-14.2.tar.xz
 a3c2b80201b89e68616f4ad30bc66aee4927c3ce50e33929ca819d5c43538898  gmp-6.3.0.tar.xz
-8f74213b56238c85a50a5329f77e06198771e70dd9a739779f4c02f65d971313  libiconv-1.17.tar.gz
-dd16fb1d67bfab79a72f5e8390735c49e3e8e70b4945a15ab1f81ddb78658fb3  make-4.4.1.tar.gz
 cc41898aac4b6e8dd5cffd7331b9d9515b912df4420a3a612b5ea2955bbeed2f  mingw-w64-v12.0.0.tar.bz2
 ab642492f5cf882b74aa0cb730cd410a81edcdbec895183ce930e706c1c759b8  mpc-1.3.1.tar.gz
 277807353a6726978996945af13e52829e3abd7a9a5b7fb2793894e18f1fcbb2  mpfr-4.2.1.tar.xz
-3caf6729c1073bf96629b57cee31eeb54f4f8129b01902c73428836550b30a3f  nasm-2.15.05.tar.xz
-590dbe0f5835f66992df096d3602d0271103f90cf8557a5d124f693c2b40d7ec  PDCurses-3.9.tar.gz
-a6456bc154999d83d0c20d968ac7ba6e7df0d02f3cb6427fb248660bacfb336e  vim-9.0.tar.bz2
